### PR TITLE
Update jetpack extension tests to work with update to @wordpress dependencies

### PR DESCRIPTION
**Pull request by glendaviesnz**
#### Changes proposed in this Pull Request:
* Changes to some Jetpack block test expectations to work with the newer Gutenberg @wordpress dependencies

Related to #19328

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
* Check our PR and run `pnpm install`
* Run `pnpm test-extensions` in the `projects/plugins/jetpack` dir and make sure all pass

### CODE REVIEW

**Review by ramonjd — APPROVED** [review]
The tests pass for me. 

I couldn't test the block changes to `projects/plugins/jetpack/extensions/blocks/recurring-payments/controls.js` because I couldn't get this branch to build. (It kept complaining about `global.window` being undefined in webpack in [locale/index.js](https://github.com/Automattic/jetpack/blob/6c7323b1ecb76efda265aa58f4a3a7c95b8cad7f/projects/plugins/jetpack/_inc/client/lib/locale/index.js#L46)

So I tested the change `onChange={ id => setMembershipAmount( id ) }` on the master branch for the Recurring payment plan controls.

It works as expected. I can add/change the payment plan for the selected payment block:


Screen Shot 2021-07-12 at 10 22 24 am

**Review by anomiex — COMMENTED** [review]
I see the several other failures that need fixing:

* Looks like the upgrade of `@wordpress/eslint-plugin` is making it complain about stuff in the e2e directory. #20333 should fix that.
* `pnpm build-search` is [failing](https://github.com/Automattic/jetpack/runs/3041985925?check_suite_focus=true#step:11:6508) with
  ```
  [3] [webpack-cli] Failed to load '/home/runner/work/jetpack/jetpack/projects/plugins/jetpack/tools/webpack.config.search.js' config
  [3] [webpack-cli] Error: Cannot find module '@wordpress/dependency-extraction-webpack-plugin/util'
  ```
  I'm guessing that `@wordpress/dependency-extraction-webpack-plugin/util` was renamed or removed?
* `pnpm build-production-client` is [failing](https://github.com/Automattic/jetpack/runs/3041985925?check_suite_focus=true#step:11:6567) with
  ```
  [build-production-client] Message:
  [build-production-client]     Module parse failed: Unexpected token (47:304)
  ```
* `pnpm build-extensions` is [failing](https://github.com/Automattic/jetpack/runs/3041985925?check_suite_focus=true#step:11:7282) with several errors about `export 'isRTL' was not found in '@wordpress/i18n'` and one about `TypeError: document.querySelectorAll is not a function`.

File: projects/plugins/jetpack/extensions/blocks/button/test/controls.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /**
```
Comment (glendaviesnz) [inline projects/plugins/jetpack/extensions/blocks/button/test/controls.js:1]: This file had a mix of tab and space indentation - not sure why that wasn't picked up by linting when it was first added. The non whitespace related change is 154-157

File: projects/packages/backup/changelog/update-internal-dependency [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Significance: patch
```
[the remark this answers is not in the document]
Comment (glendaviesnz) [reply projects/packages/backup/changelog/update-internal-dependency]: fixed - also removed some duplicate changelog entries - seems like the renovate bot added some of its own somehow

File: projects/plugins/jetpack/extensions/blocks/button/test/controls.js [hunk 3]
```diff
@@ -151,10 +151,11 @@ describe( 'Inspector settings', () => {
             render( <ButtonControls { ...defaultProps } /> );
 
             const borderPanel = screen.getByText( 'Border Settings' ).closest( 'div' );
-            const input = within( borderPanel ).getAllByLabelText( 'Border radius' )[ 1 ];
+            const input = borderPanel.querySelector( 'input[type="number"]' );
+            input.focus();
+			fireEvent.change( input, { target: { value: '6' } } );
```
[the remark this answers is not in the document]
Comment (glendaviesnz) [reply projects/plugins/jetpack/extensions/blocks/button/test/controls.js]: >What's going on with the indentation here?

Some this file had a mix of spaces and tabs - not sure why linting didnt pick that up when first committed - should be fixed now, but adds a bit of noise to the PR.

> what's the difference between using userEvent.type() and fireEvent.change() 

userEvent is the recommended approach as it tries to do a better job of replicating the range of events that happen when a user interacts with an element. fireEvent is a bit lower level and fires only the event specified. In some situations userEvent doesn't seem to work, and gutenberg have changed something about that element that stopped userEvent working, wasn't able to work out what, and check and the core team are using fireEvent to test this component so juct copied their approach.

**Comment by glendaviesnz** [comment]
> I see the several other failures that need fixing:

I was thinking that this PR would just cover getting the `test-extensions` task running, and that other build failures, etc. would be covered in other PRs, or the main PR, but happy to extend the scope of this one to cover any js related tasks in the jetpack plugin if you think that is a better approach.

File: projects/plugins/jetpack/extensions/blocks/button/test/controls.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (anomiex) [reply projects/plugins/jetpack/extensions/blocks/button/test/controls.js]: Thanks for the explanation. (:

**Comment by anomiex** [comment]
I'm ok with that approach, as long as the other PRs actually get created. I'd hate for the work you put in here to never wind up making it into master because the other parts didn't happen.

**Comment by glendaviesnz** [comment]
@anomiex I am stuck on this error:

`ERROR in.../node_modules/.pnpm/@wordpress+components@14.1.5_react-dom@16.14.0+react@16.14.0/node_modules/@wordpress/components/build-module/toolbar/toolbar-container.js 30:9-14
    "export 'isRTL' was not found in '@wordpress/i18n'`

That `pnpm build` is throwing. If I check the  `@wordpress/i18n` module in the `projects/plugins/jetpack/node_modules` folder it seems to be export `isRTL` as a method to me. Do you have any thoughts from a `pnpm` setup perspective? 

I even tried deleting the pnpm global store and re-installing everything from scratch but it didn't help

**Comment by glendaviesnz** [comment]
Also no luck so far with `ERROR in TypeError: document.querySelectorAll is not a function
        at createCache (webpack:////...jetpack/node_modules/.pnpm/@emotion+cache@10.0.29/node_modules/@emotion/cache/dist/cache.browser.esm.js?:110:26)` error on `pnpm build` in jetpack plugin - tried installed the `@emotion` babel plugin which Gutenberg repo has added , but didn't help

**Comment by anomiex** [comment]
@glendaviesnz Looks like the isRTL thing is nothing to do with pnpm. 

It seems that webpack is being configured to replace `@wordpress/i18n` with a custom implementation:
https://github.com/Automattic/jetpack/blob/5567faca99ecc793881bde9e5ec1dccd8af5f815/projects/plugins/jetpack/tools/webpack.config.extensions.js#L171-L174

And [extensions/shared/i18n-to-php.js](https://github.com/Automattic/jetpack/blob/5567faca99ecc793881bde9e5ec1dccd8af5f815/projects/plugins/jetpack/extensions/shared/i18n-to-php.js) does not export an `isRTL`. I see that was added in #13070 and hasn't been touched since.

**Comment by anomiex** [comment]
Looks like for the `document.querySelectorAll`, it's because it's using "cache.browser.esm.js" which assumes it's running in a browser. There's a "cache.esm.js" that has a check, but it's not picking that one.

Maybe the thing to do would be to further override `mainFields` at
https://github.com/Automattic/jetpack/blob/5567faca99ecc793881bde9e5ec1dccd8af5f815/projects/plugins/jetpack/tools/webpack.config.extensions.js#L149
and
https://github.com/Automattic/jetpack/blob/5567faca99ecc793881bde9e5ec1dccd8af5f815/projects/plugins/jetpack/tools/webpack.config.extensions.js#L167
to something specific instead of webpack's default that begins with "browser"? Or set a non-browser `target` so webpack uses a different default in the first place?

I wonder if that would also let us clean up some of the mess at
https://github.com/Automattic/jetpack/blob/5567faca99ecc793881bde9e5ec1dccd8af5f815/projects/plugins/jetpack/tools/webpack.config.extensions.js#L180-L210

**Comment by glendaviesnz** [comment]
@anomiex thanks for looking at this - those suggestions make sense. FYI, I probably won't have chance to follow up in this until tomorrow or Monday.

**Comment by glendaviesnz** [comment]
@fullofcaffeine your fix for isRTL looks good to me. I have also pushed [a fix for a scss issue](https://github.com/Automattic/jetpack/pull/20326/commits/26a31e716c6ff0d711b5fa46bb4c60f5f75c31c5) that showed up after the isRTL issue was gone.

I haven't had any joy with the emotion issue as mocking the missing `document.querySelectorAll` just causes flow on errors as emotion seems to be dependent in a valid return here. I tried changing the webpack `resolve: { mainFields: undefined}` to `resolve: { mainFields: ['browser', 'module', 'main']}`, and also added `@emotion/babel-plugin` to the babel config, but that didn't help.

@sarayourfriend  - when trying to update wordpress/components we are getting the following `@emotion` related error. 

```
ERROR in TypeError: document.querySelectorAll is not a function
        at createCache (webpack:////Users/glendavies/dev/a8c-projects/jetpack/node_modules/.pnpm/@emotion+cache@10.0.29/node_modules/@emotion/cache/dist/cache.browser.esm.js?:110:26)
```
Do you know if there is a babel/webpack module or config that needs adding to get builds that are using the new `@emotion` related components to run?

**Comment by jsnajdr** [comment]
The `document.querySelectorAll` error is triggered by the `StaticSiteGeneratorPlugin` webpack plugin. That plugin takes the compiled JS asset source and evaluates it inside a Node.js context to get a static markup.

The compiled JS asset is compiled for the browser, with browser versions of Emotion and other modules. And it's executed by the plugin in a Node.js environment, where the global object is supplied by the `globals` option in the webpack config. That's why references to `document` and `window` work.

But for Emotion, the existing `document` mock is not enough. In addition to `document.querySelectorAll`, Emotion uses methods like `createElement`, `appendChild`, `insertBefore`, ... I'd have to mock most of the DOM API to make it work.

A good solution would be to use `JSDOM` to mock the globals. Just pass `globals: new JSDOM()` to the `StaticSiteGeneratorPlugin` and then Emotion should have a complete DOM mock where it can add elements. There's no effect on the final static output of course. If there were supposed to be any Emotion styles in the static output, they won't be there. But I don't know enough about what exactly the static generator plugin is used for.

**Comment by anomiex** [comment]
> @fullofcaffeine your fix for isRTL looks good to me.

Once we fix the `document.querySelectorAll` thing, it doesn't seem to work after all:

```
    ERROR in RangeError: Maximum call stack size exceeded
        at Module.eval [as isRTL] (webpack:///./extensions/shared/i18n-to-php.js?:11:96)
        at Module.eval [as isRTL] (webpack:///./extensions/shared/i18n-to-php.js?:11:147)
        at Module.eval [as isRTL] (webpack:///./extensions/shared/i18n-to-php.js?:11:147)
        at Module.eval [as isRTL] (webpack:///./extensions/shared/i18n-to-php.js?:11:147)
        at Module.eval [as isRTL] (webpack:///./extensions/shared/i18n-to-php.js?:11:147)
        at Module.eval [as isRTL] (webpack:///./extensions/shared/i18n-to-php.js?:11:147)
        at Module.eval [as isRTL] (webpack:///./extensions/shared/i18n-to-php.js?:11:147)
        at Module.eval [as isRTL] (webpack:///./extensions/shared/i18n-to-php.js?:11:147)
        at Module.eval [as isRTL] (webpack:///./extensions/shared/i18n-to-php.js?:11:147)
        at Module.eval [as isRTL] (webpack:///./extensions/shared/i18n-to-php.js?:11:147)
```

I don't know if this is the best fix, but it compiles if I add another replacement
```js
            new webpack.NormalModuleReplacementPlugin(
                /^@wordpress\/i18n\/replaced$/,
                require.resolve( '@wordpress/i18n' )
            ),
```
so it can refer to `@wordpress/i18n/replaced` at
https://github.com/Automattic/jetpack/blob/26a31e716c6ff0d711b5fa46bb4c60f5f75c31c5/projects/plugins/jetpack/extensions/shared/i18n-to-php.js#L71

I have no idea whether the resulting output is actually correct though, or if it'll always just pick one directionality.

> I haven't had any joy with the emotion issue as mocking the missing `document.querySelectorAll` just causes flow on errors as emotion seems to be dependent in a valid return here. I tried changing the webpack `resolve: { mainFields: undefined}` to `resolve: { mainFields: ['module', 'main']}`, and also added `target: node` to webpack config, and `@emotion/babel-plugin` to the babel config, but neither helped.

Adding `target: node` changed it to include the `cache.esm.js` instead of `cache.browser.esm.js`, but it still fails with the same error. Looks like that's because its `isBrowser` check is for `typeof document !== 'undefined'` and we're mocking `document` so it's not undefined. But removing the mocking of `document` makes other stuff break. Sigh.

> A good solution would be to use `JSDOM` to mock the globals. Just pass `globals: new JSDOM()` to the `StaticSiteGeneratorPlugin` and then Emotion should have a complete DOM mock where it can add elements.

That almost works, but it loses other stuff that needs mocking. It compiles if we only replace the `document` mock with `document: ( new JSDOM() ).window.document`.

> But I don't know enough about what exactly the static generator plugin is used for.

As far as I can tell, it produces the following files:

* [_inc/blocks/components.css](https://github.com/Automattic/jetpack-production/blob/master/_inc/blocks/components.css)
* [_inc/blocks/components.js](https://github.com/Automattic/jetpack-production/blob/master/_inc/blocks/components.js)
* [_inc/blocks/components.rtl.css](https://github.com/Automattic/jetpack-production/blob/master/_inc/blocks/components.rtl.css)
* [_inc/blocks/frontend-nudge.html](https://github.com/Automattic/jetpack-production/blob/master/_inc/blocks/frontend-nudge.html)
* [_inc/blocks/upgrade-nudge.html](https://github.com/Automattic/jetpack-production/blob/master/_inc/blocks/upgrade-nudge.html)

I could easily be missing it, but I don't see that `components.js` being loaded anywhere. The rest are loaded from [_inc/lib/components.php](https://github.com/Automattic/jetpack-production/blob/master/_inc/lib/components.php).

**Comment by glendaviesnz** [comment]
@anomiex, not sure if it is better than your additional replacement or not, but added a custom function that excludes the `extensions/shared/i18n-to-php`  file from its [own remapping ](https://github.com/Automattic/jetpack/pull/20326/commits/8d4573157560ace6a7f7c097d43be183a3b3f212#diff-fea4386d9dcd7b31d1f58106632cded35b83c8afbca6e2783e26dfe3c5427b9cR174) - seems to work.

Also pushed the change to add in the jsdom mapping, and a fix for a changed module path in the search webpack build.

**Comment by glendaviesnz** [comment]
@anomiex, @fullofcaffeine  - it seems to successfully build now with the latest two commits that I pushed. Let me know if you still get any errors after pulling the latest

**Comment by anomiex** [comment]
> @anomiex, not sure if it is better than your additional replacement or not, but added a custom function that excludes the `extensions/shared/i18n-to-php` file from its [own remapping ](https://github.com/Automattic/jetpack/pull/20326/commits/8d4573157560ace6a7f7c097d43be183a3b3f212#diff-fea4386d9dcd7b31d1f58106632cded35b83c8afbca6e2783e26dfe3c5427b9cR174) - seems to work.

That does seem better. :+1:

**Comment by anomiex** [comment]
Sigh, hundreds of "Invalid hook call" errors. Looks like @wordpress/* now depend on react 17, so we have to update everything to that too. Working on it...

**Comment by anomiex** [comment]
Looks like the build is failing due to the check that the built JS passes ES5. It doesn't due to at least two issues:

* @wordpress/browserslist-config 4 dropped IE11 support, so Babel no longer transpiles down to ES5.
* Even if we downgrade that to 3.0.3, @wordpress/i18n isn't transpiled and includes a `const`.

We could just drop the ES5 check, or see what p9dueE-3fG-p2 comes up with.

**Comment by glendaviesnz** [comment]
Thanks @anomiex for sorting out the remaining issues!

> We could just drop the ES5 check, or see what p9dueE-3fG-p2 comes up with.

Is it an option to drop the ES5 check for this PR, and add a follow up issue for adding a bundle browser support CI check as a follow up issue? At a quick look around there are some options for checking a bundle for support at the basic [es5 versus es6 level](https://betterprogramming.pub/check-your-javascript-bundles-for-browser-support-d769c1fca4c) but not against a specific browser list, so I am guessing we would need to create something of our own for this.

**Comment by anomiex** [comment]
It looks like we could easily enough update it to an ES9 check for now. That's determined through trial and error: there's a spread operator in there somewhere that makes it fail in ES8.

The es-check tool you linked has a feature request at yowainwright/es-check#93 requesting the ability to use browserslist to determine the version. As far as I can tell, the existing eslint invocation we have as a check does the same thing that es-check would do. All we'd need with either tool is a way to turn a browserslist into the appropriate ecmaVersion.

I found [eslint-plugin-compat](https://www.npmjs.com/package/eslint-plugin-compat) that claims to check things versus the browser list, but it seems that it's about browser globals rather than ES language features. It depends on you passing the correct ecmaVersion to eslint.

**Review by anomiex — APPROVED** [review]
I'm going to go ahead and hit "approve" on this so we can merge it into the parent branch. Tests are passing and it seems at least a few people have poked at it in-browser to verify that things still work.

The failing "Changelogger use" check should pass when merged into the parent due to change entries added there.

**Comment by anomiex** [comment]
Thanks for all the work you did to help make this happen!

**Pull request merged**